File: packages/ui/src/__tests__/UIModule.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { RynBridge, MockTransport } from '@rynbridge/core';
import type { BridgeResponse } from '@rynbridge/core';
import { UIModule } from '../UIModule.js';

describe('UIModule', () => {
  let transport: MockTransport;
  let bridge: RynBridge;
  let ui: UIModule;

  beforeEach(() => {
    transport = new MockTransport();
    bridge = new RynBridge({ timeout: 5000 }, transport);
    ui = new UIModule(bridge);
  });

  function respondSuccess(payload: Record<string, unknown> = {}) {
    const sent = JSON.parse(transport.sent[transport.sent.length - 1]);
    const response: BridgeResponse = {
      id: sent.id,
      status: 'success',
      payload,
      error: null,
    };
    transport.simulateIncoming(JSON.stringify(response));
  }

  function simulateNativeEvent(module: string, action: string, payload: Record<string, unknown>) {
    const event = JSON.stringify({
      id: 'event-' + Math.random().toString(36).slice(2),
      module,
      action,
      payload,
      version: '1.0.0',
    });
    transport.simulateIncoming(event);
  }

  describe('showAlert', () => {
    it('sends correct module, action, and payload', async () => {
      const promise = ui.showAlert({ title: 'Warning', message: 'Are you sure?' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('ui');
      expect(sent.action).toBe('showAlert');
      expect(sent.payload).toEqual({ title: 'Warning', message: 'Are you sure?' });

      respondSuccess();
      await promise;
    });

    it('sends optional buttonText', async () => {
      const promise = ui.showAlert({ title: 'Info', message: 'Done', buttonText: 'OK' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.payload).toEqual({ title: 'Info', message: 'Done', buttonText: 'OK' });

      respondSuccess();
      await promise;
    });
  });

  describe('showConfirm', () => {
    it('returns true when confirmed', async () => {
      const promise = ui.showConfirm({ title: 'Delete', message: 'Delete this item?' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('ui');
      expect(sent.action).toBe('showConfirm');
      expect(sent.payload).toEqual({ title: 'Delete', message: 'Delete this item?' });

      respondSuccess({ confirmed: true });

      const result = await promise;
      expect(result).toBe(true);
    });

    it('returns false when cancelled', async () => {
      const promise = ui.showConfirm({ title: 'Delete', message: 'Delete?' });
      respondSuccess({ confirmed: false });

      const result = await promise;
      expect(result).toBe(false);
    });
  });

  describe('showToast', () => {
    it('sends fire-and-forget with message', () => {
      ui.showToast({ message: 'Saved!' });

      expect(transport.sent).toHaveLength(1);
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('ui');
      expect(sent.action).toBe('showToast');
      expect(sent.payload).toEqual({ message: 'Saved!' });
    });

    it('sends optional duration', () => {
      ui.showToast({ message: 'Loading...', duration: 3000 });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.payload).toEqual({ message: 'Loading...', duration: 3000 });
    });
  });

  describe('showActionSheet', () => {
    it('returns selected index', async () => {
      const promise = ui.showActionSheet({
        title: 'Choose action',
        options: ['Edit', 'Delete', 'Cancel'],
      });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('ui');
      expect(sent.action).toBe('showActionSheet');
      expect(sent.payload).toEqual({
        title: 'Choose action',
        options: ['Edit', 'Delete', 'Cancel'],
      });

      respondSuccess({ selectedIndex: 1 });

      const result = await promise;
      expect(result).toBe(1);
    });

    it('works without title', async () => {
      const promise = ui.showActionSheet({ options: ['A', 'B'] });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.payload).toEqual({ options: ['A', 'B'] });

      respondSuccess({ selectedIndex: 0 });

      const result = await promise;
      expect(result).toBe(0);
    });
  });

  describe('setStatusBar', () => {
    it('sends correct payload with style', async () => {
      const promise = ui.setStatusBar({ style: 'light' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('ui');
      expect(sent.action).toBe('setStatusBar');
      expect(sent.payload).toEqual({ style: 'light' });

      respondSuccess();
      await promise;
    });

    it('sends correct payload with hidden', async () => {
      const promise = ui.setStatusBar({ hidden: true });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.payload).toEqual({ hidden: true });

      respondSuccess();
      await promise;
    });
  });

  describe('showKeyboard', () => {
    it('sends fire-and-forget', () => {
      ui.showKeyboard();

      expect(transport.sent).toHaveLength(1);
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('ui');
      expect(sent.action).toBe('showKeyboard');
      expect(sent.payload).toEqual({});
    });
  });

  describe('hideKeyboard', () => {
    it('sends fire-and-forget', () => {
      ui.hideKeyboard();

      expect(transport.sent).toHaveLength(1);
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('ui');
      expect(sent.action).toBe('hideKeyboard');
      expect(sent.payload).toEqual({});
    });
  });

  describe('getKeyboardHeight', () => {
    it('returns keyboard height and visibility', async () => {
      const promise = ui.getKeyboardHeight();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('ui');
      expect(sent.action).toBe('getKeyboardHeight');
      expect(sent.payload).toEqual({});

      respondSuccess({ height: 336, visible: true });

      const result = await promise;
      expect(result).toEqual({ height: 336, visible: true });
    });

    it('returns zero height when keyboard is hidden', async () => {
      const promise = ui.getKeyboardHeight();
      respondSuccess({ height: 0, visible: false });

      const result = await promise;
      expect(result).toEqual({ height: 0, visible: false });
    });
  });

  describe('onKeyboardChange', () => {
    it('subscribes to keyboard events and returns unsubscribe function', () => {
      const received: unknown[] = [];
      const unsubscribe = ui.onKeyboardChange((data) => {
        received.push(data);
      });

      simulateNativeEvent('ui', 'keyboardChange', { height: 336, visible: true });

      expect(received).toHaveLength(1);
      expect(received[0]).toEqual({ height: 336, visible: true });

      unsubscribe();
      simulateNativeEvent('ui', 'keyboardChange', { height: 0, visible: false });

      expect(received).toHaveLength(1);
    });
  });

  describe('error handling', () => {
    it('propagates bridge errors', async () => {
      const promise = ui.showAlert({ title: 'Test', message: 'Test' });

      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'UI not available' },
      };
      transport.simulateIncoming(JSON.stringify(response));

      await expect(promise).rejects.toThrow('UI not available');
    });
  });
});

File: android/ui/src/test/kotlin/com/devryner/rynbridge/ui/UIModuleTest.kt
package com.devryner.rynbridge.ui

import com.devryner.rynbridge.core.*
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class UIModuleTest {

    @Test
    fun `showAlert`() = runTest {
        val provider = MockUIProvider()
        val module = UIModule(provider)
        val handler = module.actions["showAlert"]!!

        val result = handler(mapOf(
            "title" to BridgeValue.string("Hello"),
            "message" to BridgeValue.string("World"),
            "buttonText" to BridgeValue.string("OK")
        ))
        assertTrue(result.isEmpty())
        assertEquals("Hello", provider.lastAlertTitle)
        assertEquals("World", provider.lastAlertMessage)
        assertEquals("OK", provider.lastAlertButtonText)
    }

    @Test
    fun `showAlert defaults`() = runTest {
        val provider = MockUIProvider()
        val module = UIModule(provider)
        val handler = module.actions["showAlert"]!!

        handler(mapOf("title" to BridgeValue.string("T"), "message" to BridgeValue.string("M")))
        assertEquals("OK", provider.lastAlertButtonText)
    }

    @Test
    fun `showConfirm`() = runTest {
        val provider = MockUIProvider()
        provider.confirmResult = true
        val module = UIModule(provider)
        val handler = module.actions["showConfirm"]!!

        val result = handler(mapOf(
            "title" to BridgeValue.string("Delete?"),
            "message" to BridgeValue.string("Are you sure?"),
            "confirmText" to BridgeValue.string("Yes"),
            "cancelText" to BridgeValue.string("No")
        ))
        assertEquals(true, result["confirmed"]?.boolValue)
        assertEquals("Delete?", provider.lastConfirmTitle)
    }

    @Test
    fun `showConfirm defaults`() = runTest {
        val provider = MockUIProvider()
        val module = UIModule(provider)
        val handler = module.actions["showConfirm"]!!

        handler(mapOf("title" to BridgeValue.string("T"), "message" to BridgeValue.string("M")))
        assertEquals("Confirm", provider.lastConfirmConfirmText)
        assertEquals("Cancel", provider.lastConfirmCancelText)
    }

    @Test
    fun `showToast`() = runTest {
        val provider = MockUIProvider()
        val module = UIModule(provider)
        val handler = module.actions["showToast"]!!

        val result = handler(mapOf("message" to BridgeValue.string("Saved!"), "duration" to BridgeValue.double(3.0)))
        assertTrue(result.isEmpty())
        assertEquals("Saved!", provider.lastToastMessage)
        assertEquals(3.0, provider.lastToastDuration)
    }

    @Test
    fun `showToast default duration`() = runTest {
        val provider = MockUIProvider()
        val module = UIModule(provider)
        val handler = module.actions["showToast"]!!

        handler(mapOf("message" to BridgeValue.string("Hi")))
        assertEquals(2.0, provider.lastToastDuration)
    }

    @Test
    fun `showActionSheet`() = runTest {
        val provider = MockUIProvider()
        provider.actionSheetResult = 1
        val module = UIModule(provider)
        val handler = module.actions["showActionSheet"]!!

        val result = handler(mapOf(
            "title" to BridgeValue.string("Choose"),
            "options" to BridgeValue.array(listOf(
                BridgeValue.string("A"),
                BridgeValue.string("B"),
                BridgeValue.string("C")
            ))
        ))
        assertEquals(1L, result["selectedIndex"]?.intValue)
        assertEquals(listOf("A", "B", "C"), provider.lastActionSheetOptions)
    }

    @Test
    fun `setStatusBar`() = runTest {
        val provider = MockUIProvider()
        val module = UIModule(provider)
        val handler = module.actions["setStatusBar"]!!

        val result = handler(mapOf("style" to BridgeValue.string("light"), "hidden" to BridgeValue.bool(true)))
        assertTrue(result.isEmpty())
        assertEquals("light", provider.lastStatusBarStyle)
        assertEquals(true, provider.lastStatusBarHidden)
    }

    @Test
    fun `module name and version`() {
        val provider = MockUIProvider()
        val module = UIModule(provider)
        assertEquals("ui", module.name)
        assertEquals("0.1.0", module.version)
    }

    @Test
    fun `end to end with bridge`() = runTest {
        val transport = MockTransport()
        val bridge = RynBridge(transport = transport, config = BridgeConfig(timeout = 5000L))
        val provider = MockUIProvider()
        bridge.register(UIModule(provider))

        val requestJSON = """{"id":"req-1","module":"ui","action":"showToast","payload":{"message":"Hello"},"version":"0.1.0"}"""
        transport.simulateIncoming(requestJSON)
        transport.awaitSent(1)

        assertEquals(1, transport.sent.size)
        val json = Json { ignoreUnknownKeys = true }
        val response = json.decodeFromString<BridgeResponse>(transport.sent[0])
        assertEquals(ResponseStatus.success, response.status)

        bridge.dispose()
    }
}

private class MockUIProvider : UIProvider {
    var lastAlertTitle: String? = null
    var lastAlertMessage: String? = null
    var lastAlertButtonText: String? = null
    var lastConfirmTitle: String? = null
    var lastConfirmConfirmText: String? = null
    var lastConfirmCancelText: String? = null
    var confirmResult = false
    var lastToastMessage: String? = null
    var lastToastDuration: Double? = null
    var lastActionSheetTitle: String? = null
    var lastActionSheetOptions: List<String>? = null
    var actionSheetResult = 0
    var lastStatusBarStyle: String? = null
    var lastStatusBarHidden: Boolean? = null

    override suspend fun showAlert(title: String, message: String, buttonText: String) {
        lastAlertTitle = title
        lastAlertMessage = message
        lastAlertButtonText = buttonText
    }

    override suspend fun showConfirm(title: String, message: String, confirmText: String, cancelText: String): Boolean {
        lastConfirmTitle = title
        lastConfirmConfirmText = confirmText
        lastConfirmCancelText = cancelText
        return confirmResult
    }

    override fun showToast(message: String, duration: Double) {
        lastToastMessage = message
        lastToastDuration = duration
    }

    override suspend fun showActionSheet(title: String?, options: List<String>): Int {
        lastActionSheetTitle = title
        lastActionSheetOptions = options
        return actionSheetResult
    }

    override suspend fun setStatusBar(style: String?, hidden: Boolean?) {
        lastStatusBarStyle = style
        lastStatusBarHidden = hidden
    }

    override fun showKeyboard() {}
    override fun hideKeyboard() {}
    override suspend fun getKeyboardHeight(): KeyboardInfo = KeyboardInfo(height = 0.0, visible = false)
}
